src/lcm/grids.py

```diff
@@ -1,59 +1,111 @@
 """Functions to generate and work with different kinds of grids.
 
-Grid generation functions have the arguments:
+Grid generation functions must have the following signature:
 
-- start
-- stop
-- n_points
+    Signature (start: Scalar, stop: Scalar, n_points: int) -> jax.Array
 
-interpolation info functions have the arguments
+They take start and end points and create a grid of points between them.
 
-- value
-- start
-- stop
-- n_points
+
+Interpolation info functions must have the following signature:
+
+    Signature (
+        value: Scalar,
+        start: Scalar,
+        stop: Scalar,
+        n_points: int
+    ) -> Scalar
+
+They take the information required to generate a grid, and return an index corresponding
+to the value, which is a point in the space but not necessarily a grid point.
 
 Some of the arguments will not be used by all functions but the aligned interface makes
 it easy to call functions interchangeably.
 
 """
 
 import jax.numpy as jnp
+from jax import Array
+
+from lcm.typing import Scalar
+
+
+def linspace(start: Scalar, stop: Scalar, n_points: int) -> Array:
+    """Wrapper around jnp.linspace.
 
+    Returns a linearly spaced grid between start and stop with n_points, including both
+    endpoints.
 
-def linspace(start, stop, n_points):
+    """
     return jnp.linspace(start, stop, n_points)
 
 
-def get_linspace_coordinate(value, start, stop, n_points):
-    """Map a value into the input needed for map_coordinates."""
+def get_linspace_coordinate(
+    value: Scalar,
+    start: Scalar,
+    stop: Scalar,
+    n_points: int,
+) -> Scalar:
+    """Map a value into the input needed for jax.scipy.ndimage.map_coordinates."""
```

Comment:
1. If the value is outside the grid, the functions in `grids.py` will return a coordinate smaller than 0 or larger than the last index (at least in some dimension). In my opinion, the behavior of the interpolator, in this case, should not be handled in `grids.py`. The interpolator is defined and used in the `function_evaluator.py` module (look for the `_get_interpolator` function). Here we set the option `mode="nearest"`. This will choose the closest point in the grid (i.e., in the 1D case, if the coordinate is -1, it will return the value at the coordinate 0). I will clean up the `function_evaluator.py` module soon, in which case this will be better documented. For now, I've added an illustrative test to `test_grids.py`.

2. That is a good question. I will reply to you about this and add the result to the developer notes.

Reply:
Re 1. — I think we'll need to discuss that once more. Taking the example of consumption, I would think that "nearest" means that its marginal utility is zero outside the grid, which will lead to very strange behaviour if the grid has been chosen poorly. Will be very hard to "debug" then (no bug in the strict sense, ofc).

My guess would be that "reflect" may do what I would think is the best approximation (take the two extreme gridpoints and extrapolate).

Maybe we could expose that kwarg instead of hard-coding it in `_get_interpolator`? In any case, we should use whatever is available in the  interpolation function we'll end up using.

Not necessarily for this PR, feel free to open an issue instead.

Reply:
Re 1. - These are good points. I've opened an issue task regarding this. Will be worked on once I tackle the interpolation functionality in LCM.

Re 2. - I've also added a task to that issue to compare the feature set of RegularGridInterpolator to the strategy we are using now. First tests show that it may be slower, and there is some confusion on whether it works with log-grids. The other scipy interpolation functions are not implemented in JAX.